### ISSUE TITLE
Bug in magic transformer ?

### ISSUE THREAD

**Issue by rhendrix42** [souffle-lang/souffle#1679]
Hi guys, 
Consider the following program:
```
.decl oAhz(A:float, B:float)
.decl AvvH(A:float)
.decl BMaq(A:float, B:float)
.decl dnIW(A:float, B:float) 
.decl wVSI(A:float) 
.decl XxRQ(A:float, B:float, C:float) 
.decl uAsC(A:float) 
.decl ByPj(A:float, B:float) 
.decl KHUU(A:float) 
.decl fefn(A:float, B:float)
.decl aZKS(A:float) 
.decl KrUb(A:float, B:float)
.decl MBkI(A:float)
.decl jicu(A:float)
.decl Sftg(A:float) 
.decl vfkR(A:float) 
.decl PqGx(A:float, B:float, C:float)
.decl OExx(A:float)
.decl KUsh(A:float) 
.decl KmRO(A:float, B:float, C:float) 
.decl YkdK(A:float) 
.decl QUlY(A:float)
.decl negate(A:float, B:float) 

.input oAhz
.input AvvH
.input BMaq

XxRQ(w,s,E) :- AvvH(Q), AvvH(s), AvvH(E), AvvH(s), oAhz(w,F).   
uAsC(max(-x,-10.605)) :- BMaq(x,x), BMaq(x,x), x <= x.   
ByPj(z,x) :- XxRQ(y,z,x), ord(x) >= ord(z).   
ByPj(Q,min(q,9.585)) :- uAsC(q), oAhz(E,q), BMaq(Q,E), uAsC(Q), uAsC(M), 5.561 = q.   
KHUU(R) :- ByPj(R,T).   
fefn(o,min(l,O)) :- ByPj(l,o), BMaq(l,N), ByPj(m,m), oAhz(W,O), ord(W) <= ord(N).   
aZKS(g) :- ByPj(g,M), ord(M) > ord(g).   
KrUb(max(J,M),v) :- XxRQ(v,M,J), BMaq(M,M), AvvH(h), ByPj(h,V), ByPj(M,h).   
MBkI(min(--i,i)) :- AvvH(T), ByPj(i,T), -2.274 > i.   
jicu(min(U,v)) :- fefn(v,U), ByPj(v,U).   
Sftg(y) :- MBkI(y), !fefn(y,y).   
vfkR(F) :- aZKS(F), Sftg(F), !aZKS(F).   
PqGx(m,f,g) :- jicu(b), vfkR(m), KrUb(D,g), KrUb(g,f), !Sftg(b).   
OExx(r) :- KHUU(F), aZKS(r), jicu(r), MBkI(r), vfkR(F), !KHUU(F).   
KUsh(w), wVSI(J) :- XxRQ(w,J,w), Sftg(w), KHUU(w), !vfkR(w), w = w.   
KmRO(D,V,max(--r,V)), dnIW(D,D) :- PqGx(D,T,r), OExx(X), AvvH(k), Sftg(V), oAhz(D,k), !jicu(r), ord(V) != ord(r).   
YkdK(min(--i,V)), dnIW(V,i):- KrUb(O,i), Sftg(V), KUsh(V), !KHUU(O), V < i.   
negate(L,L) :- Sftg(Y), vfkR(h), MBkI(L), !dnIW(L,L).
QUlY(h) :- Sftg(Y), vfkR(h), MBkI(L), !negate(L,L).   
 

.output QUlY
```
Enabling magic transformation changes the result.

```
$ souffle -w min.dl
$ cat QUlY.csv
$ souffle -w --magic-transform=* min.dl 
$ cat QUlY.csv
-3.68000007
-3.97399998
-4.33300018
-5.52199984
-5.80700016
-7.27400017
-7.43599987
-7.72200012
-7.98000002
-8.1960001
-13.9949999
-14.1680002
-14.2299995
-14.5179996
-16.7490005
-18.5279999
-20.2199993
-20.3619995
-20.7450008
-21.6119995
-21.6359997
-22.5300007
-23.5510006
-23.8570004
-25.6660004
-25.8290005
```
The result should remain the same, right ?

Fact Files: 
[AvvH.facts](https://github.com/souffle-lang/souffle/files/5312265/AvvH.facts.zip)
[BMaq.facts](https://github.com/souffle-lang/souffle/files/5312266/BMaq.facts.zip)
[oAhz.facts](https://github.com/souffle-lang/souffle/files/5312267/oAhz.facts.zip)

I am running this on 64-bit ubuntu16.04
The revision i am using is : 2995ff775ccb23342b06b457835662b03c5bc1f5

**Comment by azreika** [comment]
This will need to be investigated, but there is a chance that this is because the program uses float equalities (e.g. `w = x`). The transformation should be checked manually just in case.

**Comment by SamArch27** [comment]
> This will need to be investigated, but there is a chance that this is because the program uses float equalities (e.g. `w = x`). The transformation should be checked manually just in case.

Yep, this is a bug for float equalities for the interpreter. I will put in a quick PR for this now.

**Comment by rhendrix42** [comment]
Hmm, now the result is just reversed
```
$ souffle -w -c --magic-transform=* min.dl
$ cat QUlY.csv 
-25.8290005
-25.6660004
-23.8570004
-23.5510006
-22.5300007
-21.6359997
-21.6119995
-20.7450008
-20.3619995
-20.2199993
-18.5279999
-16.7490005
-14.5179996
-14.2299995
-14.1680002
-13.9949999
-8.1960001
-7.98000002
-7.72200012
-7.43599987
-7.27400017
-5.80700016
-5.52199984
-4.33300018
-3.97399998
-3.68000007

```

**Comment by azreika** [comment]
Just a quick update that I'm looking into this  (manually transforming the program is a pain), but the problem could be accidentally skipped relations at the positive-labelling stage of the magic-set transformer causing a negated cycle. Still investigating.

**Comment by azreika** [comment]
After looking into it, seems like the problem is because the ignored-relation property check isn't being propagated properly in the magic-set transformation. Essentially:

* We ignore transforming certain relations to maintain semantics, such as relations tagged as `eqrel`, relations using counters, relations that aren't specified by the user, etc.

* To prevent potential float-related equality issues at the time, a check was added to ignore relations that use float-values.

* During positive-labelling, if the head of a rule is labelled, then all body atoms have to be labelled. But, an ignored-relation should not be labelled. Therefore, all relations that use ignored relations in one of their bodies should be ignored. This does not seem to be the case at the moment, as this is produced after positive-labelling currently:
```
@neglabel.jicu(@abdul0) :-
   @poscopy_1.fefn.{ff}(v,U),
   ByPj(v,U),
   @abdul0 = min(U,v).
```
`ByPj` should be poscopy'd. However, it's ignored due to the use of float-values in its clause bodies. As a result, `jicu` should not be transformed either to prevent this labelling requirement.

Still need to come up with a minimal example to make sure this is the case + a fix if it is.

Thanks for bringing up the problem!

**Comment by azreika** [comment]
Here's a very minified program with the same problem that I wrote based on the original program:
```
.decl Pairs(A:float, B:float)
Pairs(0,0).

.decl First(A:float)
.decl DupFirst(A:float, B:float)
.decl FirstAgain(A:float)
.decl Fail()

.decl Out(A:float)
.output Out

First(x) :- Pairs(x,_).

DupFirst(x,x) :- First(x), x < 100.

FirstAgain(x) :- DupFirst(x,_).

Fail() :- FirstAgain(x), !FirstAgain(x).

Out(x) :- Fail(), First(x).
```

With MST, the program outputs `0`, and nothing without. Notice that `Fail()` should always be false, so no output is possible. This is true in the original program too (see: `vfkR(F) :- aZKS(F), Sftg(F), !aZKS(F).`). I'll keep looking into what's happening tomorrow based on this example.

**Comment by azreika** [comment]
Ok, bug found. The relation `DupFirst` is not transformed because of the use of a float-operator (purposefully, to avoid float-eq-related issues). This means it is not positively-labelled in our current system. However, positive-labelling is essential in separating relation dependencies to prevent stratification issues. This issue arises in the original program (and the minified version).

To prevent this, we have two possible routes:
* Propagate the ignored-relation property to other relations to prevent the requirement for labelling, or
* Positively-label ignored relations anyway - this does not transform any rules, only copy them to others

The problem with positively-labelling ignored relations is that some rule contents may not be pure. The only example that comes to mind at the moment is the counter operator `$`. Duplicating these rules may cause semantic issues. The first option is, however, significantly more complex, and will also require looking into stopping the negative-labelling of some relations.

**Comment by azreika** [comment]
PR #1699 should fix this issue. In future, we might want to revisit the ignoring criteria altogther, and allow magic-set on relations using floating point values.

**Comment by azreika** [comment]
Fixed by PR #1699.
